File: src/agent_os_cli/lib/agent_os_cli/commands/memory.ex
defmodule AgentOS.CLI.Commands.Memory do
  @moduledoc "Memory commands: search, show."

  alias AgentOS.CLI.{HttpClient, Output}

  @doc "Routes memory subcommands to their handlers."
  def run(["search", query], opts) do
    search(query, opts)
  end

  def run(["show", id], opts) do
    show(id, opts)
  end

  def run(_, _opts) do
    Output.info("""
    Usage: agent-os memory <subcommand> [options]

    Subcommands:
      search    <query>
      show      <id>
    """)
  end

  # --- Subcommands ---

  defp search(query, opts) do
    encoded = URI.encode_www_form(query)

    case HttpClient.get("/api/v1/memory/search?q=#{encoded}", opts) do
      {:ok, %{"results" => results}} when is_list(results) ->
        if opts[:json] do
          Output.json(results)
        else
          headers = ["ID", "Type", "Content", "Score"]

          rows =
            Enum.map(results, fn r ->
              [
                Map.get(r, "id", ""),
                Map.get(r, "type", ""),
                truncate(Map.get(r, "content", ""), 60),
                to_string(Map.get(r, "score", ""))
              ]
            end)

          Output.table(headers, rows)
        end

      {:ok, body} ->
        Output.json(body)

      {:error, reason} ->
        Output.error("Failed to search memory: #{inspect(reason)}")
    end
  end

  defp show(id, opts) do
    case HttpClient.get("/api/v1/memory/#{id}", opts) do
      {:ok, body} ->
        if opts[:json] do
          Output.json(body)
        else
          headers = ["Field", "Value"]

          rows =
            Enum.map(body, fn {k, v} ->
              [to_string(k), truncate(to_string(v), 80)]
            end)

          Output.table(headers, rows)
        end

      {:error, reason} ->
        Output.error("Failed to get memory entry: #{inspect(reason)}")
    end
  end

  defp truncate(str, max) when byte_size(str) <= max, do: str
  defp truncate(str, max), do: String.slice(str, 0, max - 3) <> "..."
end

File: cli/src/commands/memory.js
/**
 * Memory commands — search, show.
 */

import * as http from '../http.js';
import * as out from '../output.js';

/**
 * Search memory.
 * GET /api/v1/memory/search?q=<query>
 * @param {string[]} args - Positional arguments (args[0] = search query)
 * @param {object} opts - Global CLI options
 */
export async function search(args, opts) {
  const query = args[0];
  if (!query) {
    out.error('Search query is required: agent-os memory search "query"');
    process.exit(1);
  }

  const encoded = encodeURIComponent(query);
  const result = await http.get(`/api/v1/memory/search?q=${encoded}`, opts);

  if (opts.json) {
    out.json(result);
    return;
  }

  const items = result.data || result.results || result;
  if (!Array.isArray(items) || items.length === 0) {
    out.info('No results found.');
    return;
  }

  const headers = ['ID', 'SCORE', 'SUMMARY'];
  const rows = items.map((item) => [
    item.id || '-',
    item.score != null ? String(item.score) : '-',
    (item.summary || item.content || '-').slice(0, 60),
  ]);
  out.table(headers, rows);
}

/**
 * Show a memory entry.
 * GET /api/v1/memory/:id
 * @param {string[]} args - Positional arguments (args[0] = memory id)
 * @param {object} opts - Global CLI options
 */
export async function show(args, opts) {
  const id = args[0];
  if (!id) {
    out.error('Memory ID is required: agent-os memory show <id>');
    process.exit(1);
  }

  const result = await http.get(`/api/v1/memory/${id}`, opts);

  if (opts.json) {
    out.json(result);
  } else {
    const entry = result.data || result;
    out.json(entry);
  }
}
